File: app/src/main/java/com/kim/kexuetuokouxiu/service/MediaNotificationManager.java
package com.kim.kexuetuokouxiu.service;

import android.app.Notification;
import android.app.PendingIntent;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.graphics.BitmapFactory;
import android.graphics.Color;
import android.os.RemoteException;
import android.support.annotation.NonNull;
import android.support.v4.app.NotificationManagerCompat;
import android.support.v4.media.MediaDescriptionCompat;
import android.support.v4.media.MediaMetadataCompat;
import android.support.v4.media.session.MediaControllerCompat;
import android.support.v4.media.session.MediaSessionCompat;
import android.support.v4.media.session.PlaybackStateCompat;
import android.support.v7.app.NotificationCompat;

import com.kim.kexuetuokouxiu.R;
import com.kim.kexuetuokouxiu.app.activity.DetailActivity;
import com.kim.kexuetuokouxiu.app.presenter.DetailPresenterImpl;
import com.kim.kexuetuokouxiu.helper.ResourceHelper;
import com.kim.kexuetuokouxiu.utils.LogUtil;

/**
 * Created by Weya on 2017/1/9.
 */

public class MediaNotificationManager extends BroadcastReceiver {

    private static final int NOTIFICATION_ID = 1994;
    private static final int REQUEST_CODE = 101;

    public static final String ACTION_PAUSE = "com.kim.kexuetuokouxiu.pause";
    public static final String ACTION_PLAY = "com.kim.kexuetuokouxiu.play";
    public static final String ACTION_PREV = "com.kim.kexuetuokouxiu.prev";
    public static final String ACTION_NEXT = "com.kim.kexuetuokouxiu.next";
    public static final String ACTION_STOP_CASTING = "com.kim.kexuetuokouxiu.stop_cast";


    private final PlayService mService;
    private MediaSessionCompat.Token mSessionToken;
    private MediaControllerCompat mController;
    private MediaControllerCompat.TransportControls mTransportControls;

    private PlaybackStateCompat mPlaybackState;
    private MediaMetadataCompat mMetadata;

    private final NotificationManagerCompat mNotificationManager;

    private final PendingIntent mPauseIntent;
    private final PendingIntent mPlayIntent;
    private final PendingIntent mPreviousIntent;
    private final PendingIntent mNextIntent;

    private final PendingIntent mStopCastIntent;

    private final int mNotificationColor;

    private boolean mStarted = false;

    public MediaNotificationManager(PlayService mService) throws RemoteException {
        this.mService = mService;

        updateSessionToken();
        mNotificationColor = ResourceHelper.getThemeColor(mService, R.attr.colorPrimary, Color.DKGRAY);
        mNotificationManager = NotificationManagerCompat.from(mService);

        String pkg = mService.getPackageName();
        mPauseIntent = PendingIntent.getBroadcast(mService, REQUEST_CODE, new Intent(ACTION_PAUSE).setPackage(pkg), PendingIntent.FLAG_CANCEL_CURRENT);
        mPlayIntent = PendingIntent.getBroadcast(mService, REQUEST_CODE, new Intent(ACTION_PLAY).setPackage(pkg), PendingIntent.FLAG_CANCEL_CURRENT);
        mPreviousIntent = PendingIntent.getBroadcast(mService, REQUEST_CODE, new Intent(ACTION_PREV).setPackage(pkg), PendingIntent.FLAG_CANCEL_CURRENT);
        mNextIntent = PendingIntent.getBroadcast(mService, REQUEST_CODE, new Intent(ACTION_NEXT).setPackage(pkg), PendingIntent.FLAG_CANCEL_CURRENT);
        mStopCastIntent = PendingIntent.getBroadcast(mService, REQUEST_CODE, new Intent(ACTION_STOP_CASTING).setPackage(pkg), PendingIntent.FLAG_CANCEL_CURRENT);

        mNotificationManager.cancelAll();
    }

    @Override
    public void onReceive(Context context, Intent intent) {
        final String action = intent.getAction();
        LogUtil.d(MediaNotificationManager.class, "Received", "intent with action " + action);
        switch (action) {
            case ACTION_PAUSE:
                mTransportControls.pause();
                break;
            case ACTION_PLAY:
                mTransportControls.play();
                break;
            case ACTION_NEXT:
                mTransportControls.skipToNext();
                break;
            case ACTION_PREV:
                mTransportControls.skipToPrevious();
                break;
            case ACTION_STOP_CASTING:
                Intent i = new Intent(context, PlayService.class);
                i.setAction(PlayService.ACTION_CMD);
                i.putExtra(PlayService.CMD_NAME, PlayService.CMD_STOP_CASTING);
                mService.startService(i);
                break;
            default:
                LogUtil.w(MediaNotificationManager.class, "Received", "Unknown intent ignored. Action=" + action);
        }
    }

    /**
     * Posts the notification and starts tracking the session to keep it
     * updated. The notification will automatically be removed if the session is
     * destroyed before {@link #stopNotification} is called.
     */
    public void startNotification() {
        if (!mStarted) {
            mMetadata = mController.getMetadata();
            mPlaybackState = mController.getPlaybackState();

            // The notification must be updated after setting started to true
            Notification notification = createNotification();
            if (notification != null) {
                mController.registerCallback(mCb);
                IntentFilter filter = new IntentFilter();
                filter.addAction(ACTION_NEXT);
                filter.addAction(ACTION_PAUSE);
                filter.addAction(ACTION_PLAY);
                filter.addAction(ACTION_PREV);
                filter.addAction(ACTION_STOP_CASTING);
                mService.registerReceiver(this, filter);

                mService.startForeground(NOTIFICATION_ID, notification);
                mStarted = true;
            }
        }
    }

    /**
     * Removes the notification and stops tracking the session. If the session
     * was destroyed this has no effect.
     */
    public void stopNotification() {
        if (mStarted) {
            mStarted = false;
            mController.unregisterCallback(mCb);
            try {
                mNotificationManager.cancel(NOTIFICATION_ID);
                mService.unregisterReceiver(this);
            } catch (IllegalArgumentException ex) {
                // ignore if the receiver is not registered.
            }
            mService.stopForeground(true);
        }
    }

    /**
     * Update the state based on a change on the session token. Called either when
     * we are running for the first time or when the media session owner has destroyed the session
     * (see {@link android.media.session.MediaController.Callback#onSessionDestroyed()})
     */
    private void updateSessionToken() throws RemoteException {
        MediaSessionCompat.Token freshToken = mService.getSessionToken();
        if (mSessionToken == null && freshToken != null ||
                mSessionToken != null && !mSessionToken.equals(freshToken)) {
            if (mController != null) {
                mController.unregisterCallback(mCb);
            }
            mSessionToken = freshToken;
            if (mSessionToken != null) {
                mController = new MediaControllerCompat(mService, mSessionToken);
                mTransportControls = mController.getTransportControls();
                if (mStarted) {
                    mController.registerCallback(mCb);
                }
            }
        }
    }

    private PendingIntent createContentIntent(MediaDescriptionCompat description) {
        Intent openUI = new Intent(mService, DetailActivity.class);
        openUI.setFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP);
        if (description != null) {
            openUI.putExtra(DetailPresenterImpl.EXTRA_CURRENT_MEDIA_DESCRIPTION, description);
        }
        return PendingIntent.getActivity(mService, REQUEST_CODE, openUI,
                PendingIntent.FLAG_CANCEL_CURRENT);
    }

    private final MediaControllerCompat.Callback mCb = new MediaControllerCompat.Callback() {
        @Override
        public void onPlaybackStateChanged(@NonNull PlaybackStateCompat state) {
            mPlaybackState = state;
            LogUtil.d(MediaNotificationManager.class, "onPlaybackStateChanged", "new playback state" + state);
            if (state.getState() == PlaybackStateCompat.STATE_STOPPED ||
                    state.getState() == PlaybackStateCompat.STATE_NONE) {
                stopNotification();
            } else {
                Notification notification = createNotification();
                if (notification != null) {
                    mNotificationManager.notify(NOTIFICATION_ID, notification);
                }
            }
        }

        @Override
        public void onMetadataChanged(MediaMetadataCompat metadata) {
            mMetadata = metadata;
            LogUtil.d(MediaNotificationManager.class, "onMetadataChanged", "new metadata " + metadata);
            Notification notification = createNotification();
            if (notification != null) {
                mNotificationManager.notify(NOTIFICATION_ID, notification);
            }
        }

        @Override
        public void onSessionDestroyed() {
            super.onSessionDestroyed();
            LogUtil.d(MediaDescriptionCompat.class, "onSessionDestroyed", "Session was destroyed, resetting to the new session token");
            try {
                updateSessionToken();
            } catch (RemoteException e) {
                LogUtil.e(MediaNotificationManager.class, "onSessionDestroyed", "could not connect media controller " + e.getMessage());
            }
        }
    };

    private Notification createNotification() {
        LogUtil.d(MediaNotificationManager.class, "createNotification", "updateNotificationMetadata. mMetadata=" + mMetadata);
        if (mMetadata == null || mPlaybackState == null) {
            return null;
        }

        NotificationCompat.Builder notificationBuilder = new NotificationCompat.Builder(mService);
        int playPauseButtonPosition = 0;

        // If skip to previous action is enabled
        if ((mPlaybackState.getActions() & PlaybackStateCompat.ACTION_SKIP_TO_PREVIOUS) != 0) {
            notificationBuilder.addAction(R.drawable.ic_skip_previous_white_24dp, "上一首", mPreviousIntent);

            // If there is a "skip to previous" button, the play/pause button will
            // be the second one. We need to keep track of it, because the MediaStyle notification
            // requires to specify the index of the buttons (actions) that should be visible
            // when in compact view.
            playPauseButtonPosition = 1;
        }

        addPlayPauseAction(notificationBuilder);

        // If skip to next action is enabled
        if ((mPlaybackState.getActions() & PlaybackStateCompat.ACTION_SKIP_TO_NEXT) != 0) {
            notificationBuilder.addAction(R.drawable.ic_skip_next_white_24dp, "下一首", mNextIntent);
        }

        MediaDescriptionCompat description = mMetadata.getDescription();

        notificationBuilder
                .setStyle(new NotificationCompat.MediaStyle()
                        .setShowActionsInCompactView(
                                new int[]{playPauseButtonPosition})  // show only play/pause in compact view
                        .setMediaSession(mSessionToken))
                .setColor(mNotificationColor)
                .setSmallIcon(R.drawable.kexuetuokouxiu_2014)
                .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
                .setUsesChronometer(true)
                .setContentIntent(createContentIntent(description))
                .setContentTitle(description.getTitle())
                .setContentText(description.getSubtitle())
                .setLargeIcon(BitmapFactory.decodeResource(mService.getResources(), R.drawable.kexuetuokouxiu_2014));

        if (mController != null && mController.getExtras() != null) {
            String castName = mController.getExtras().getString(PlayService.EXTRA_CONNECTED_CAST);
            if (castName != null) {
                String castInfo = mService.getResources()
                        .getString(R.string.casting_to_device, castName);
                notificationBuilder.setSubText(castInfo);
                notificationBuilder.addAction(R.drawable.ic_close_black_24dp,
                        "停止", mStopCastIntent);
            }
        }

        setNotificationPlaybackState(notificationBuilder);
        return notificationBuilder.build();
    }

    private void addPlayPauseAction(NotificationCompat.Builder builder) {
        LogUtil.d(MediaNotificationManager.class, "addPlayPauseAction", "updatePlayPauseAction");
        String label;
        int icon;
        PendingIntent intent;
        if (mPlaybackState.getState() == PlaybackStateCompat.STATE_PLAYING) {
            label = "暂停";
            icon = R.drawable.uamp_ic_pause_white_24dp;
            intent = mPauseIntent;
        } else {
            label = "播放";
            icon = R.drawable.uamp_ic_play_arrow_white_24dp;
            intent = mPlayIntent;
        }
        builder.addAction(new NotificationCompat.Action(icon, label, intent));
    }

    private void setNotificationPlaybackState(NotificationCompat.Builder builder) {
        LogUtil.d(MediaNotificationManager.class, "setNotificationPlaybackState", "updateNotificationPlaybackState. mPlaybackState=" + mPlaybackState);
        if (mPlaybackState == null || !mStarted) {
            LogUtil.d(MediaNotificationManager.class, "setNotificationPlaybackState", "updateNotificationPlaybackState. cancelling notification!");
            mService.stopForeground(true);
            return;
        }
        if (mPlaybackState.getState() == PlaybackStateCompat.STATE_PLAYING
                && mPlaybackState.getPosition() >= 0) {
            LogUtil.d(MediaNotificationManager.class, "setNotificationPlaybackState", "updateNotificationPlaybackState. updating playback position to " + (System.currentTimeMillis() - mPlaybackState.getPosition()) / 1000 + " seconds");
            builder
                    .setWhen(System.currentTimeMillis() - mPlaybackState.getPosition())
                    .setShowWhen(true)
                    .setUsesChronometer(true);
        } else {
            LogUtil.d(MediaNotificationManager.class, "setNotificationPlaybackState", "updateNotificationPlaybackState. hiding playback position");
            builder
                    .setWhen(0)
                    .setShowWhen(false)
                    .setUsesChronometer(false);
        }

        // Make sure that the notification can be dismissed by the user when we are not playing:
        builder.setOngoing(mPlaybackState.getState() == PlaybackStateCompat.STATE_PLAYING);
    }
}

File: app/src/main/java/com/kim/kexuetuokouxiu/MediaNotificationManager.kt
package com.kim.kexuetuokouxiu

import android.app.Notification
import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Color
import android.os.RemoteException
import android.support.v4.app.NotificationManagerCompat
import android.support.v4.media.MediaDescriptionCompat
import android.support.v4.media.MediaMetadataCompat
import android.support.v4.media.session.MediaControllerCompat
import android.support.v4.media.session.MediaSessionCompat
import android.support.v4.media.session.PlaybackStateCompat
import android.support.v7.app.NotificationCompat
import com.kim.kexuetuokouxiu.app.activity.DetailActivity
import com.kim.kexuetuokouxiu.helper.LogHelper
import com.kim.kexuetuokouxiu.helper.ResourceHelper
import com.kim.kexuetuokouxiu.service.PlayService

/**
 * Created by Weya on 2016/11/27.
 */

class MediaNotificationManager(private val mPlayService: PlayService) : BroadcastReceiver() {

    private val TAG: String = LogHelper.makeLogTag(MediaNotificationManager::class.java)

    private val NOTIFICATION_ID = 41340727
    private val REQUEST_CODE = 100

    public val ACTION_PAUSE = "com.kim.kexuetuokouxiu.pause"
    public val ACTION_PLAY = "com.kim.kexuetuokouxiu.play"
    public val ACTION_PREV = "com.kim.kexuetuokouxiu.prev"
    public val ACTION_NEXT = "com.kim.kexuetuokouxiu.next"
    public val ACTION_STOP_CASTING = "com.kim.kexuetuokouxiu.stop_cast"

    private var mSessionToken: MediaSessionCompat.Token? = null
    private var mController: MediaControllerCompat? = null
    private var mTransportControls: MediaControllerCompat.TransportControls? = null

    private var mPlaybackState: PlaybackStateCompat? = null
    private var mMetadata: MediaMetadataCompat? = null

    private val mNotificationManager: NotificationManagerCompat

    private val mPauseIntent: PendingIntent
    private val mPlayIntent: PendingIntent
    private val mPreviousIntent: PendingIntent
    private val mNextIntent: PendingIntent

    private val mStopCastIntent: PendingIntent

    private val mNotificationColor: Int

    private var mStarted = false

    init {
        updateSessionToken()

        mNotificationColor = ResourceHelper.getThemeColor(mPlayService, R.attr.colorPrimary, Color.DKGRAY)
        mNotificationManager = NotificationManagerCompat.from(mPlayService)

        val pkg: String = mPlayService.packageName
        mPauseIntent = PendingIntent.getBroadcast(mPlayService, REQUEST_CODE, Intent(ACTION_PAUSE).setPackage(pkg), PendingIntent.FLAG_CANCEL_CURRENT)
        mPlayIntent = PendingIntent.getBroadcast(mPlayService, REQUEST_CODE, Intent(ACTION_PLAY).setPackage(pkg), PendingIntent.FLAG_CANCEL_CURRENT)
        mPreviousIntent = PendingIntent.getBroadcast(mPlayService, REQUEST_CODE, Intent(ACTION_PREV).setPackage(pkg), PendingIntent.FLAG_CANCEL_CURRENT)
        mNextIntent = PendingIntent.getBroadcast(mPlayService, REQUEST_CODE, Intent(ACTION_NEXT).setPackage(pkg), PendingIntent.FLAG_CANCEL_CURRENT)
        mStopCastIntent = PendingIntent.getBroadcast(mPlayService, REQUEST_CODE, Intent(ACTION_STOP_CASTING).setPackage(pkg), PendingIntent.FLAG_CANCEL_CURRENT)

        mNotificationManager.cancelAll()
    }

    /**
     * 显示通知并跟踪Session保持通知更新。
     * 如果Session在调用{@link #stopNotification}前被销毁，通知将会自动被移除
     */
    public fun startNotification() {
        if (!mStarted) {
            mMetadata = mController?.metadata
            mPlaybackState = mController?.playbackState
            val notification: Notification? = createNotification()
            if (notification != null) {
                mController?.registerCallback(mCb)
                val filter: IntentFilter = IntentFilter()
                filter.addAction(ACTION_NEXT)
                filter.addAction(ACTION_PAUSE)
                filter.addAction(ACTION_PLAY)
                filter.addAction(ACTION_PREV)
                filter.addAction(ACTION_STOP_CASTING)
                mPlayService.registerReceiver(this, filter)

                mPlayService.startForeground(NOTIFICATION_ID, notification)
                mStarted = true
            }
        }
    }

    /**
     * 移除通知，停止跟踪Session
     * 如果Session已经被销毁，这个方法则没有影响
     */
    public fun stopNotification() {
        if (mStarted) {
            mStarted = false
            mController?.unregisterCallback(mCb)
            try {
                mNotificationManager.cancel(NOTIFICATION_ID)
                mPlayService.unregisterReceiver(this)
            } catch(e: Exception) {
            }
            mPlayService.stopForeground(true)
        }
    }

    override fun onReceive(context: Context, intent: Intent) {
        val action: String = intent.action
        LogHelper.d(TAG, "Received intent with action " + action)
        when (action) {
            ACTION_PAUSE -> mTransportControls?.pause()
            ACTION_PLAY -> mTransportControls?.play()
            ACTION_NEXT -> mTransportControls?.skipToNext()
            ACTION_PREV -> mTransportControls?.skipToPrevious()
            ACTION_STOP_CASTING -> {
                // todo
                val i = Intent(context, PlayService::class.java)
                i.action = ""
                i.putExtra("", "")
                mPlayService.startService(i)
            }
        }

    }

    /**
     * 基于Session Token更新状态
     * 在第一次运行或Media Session被所有者销毁时被调用
     */
    private fun updateSessionToken() {
        val freshToken = mPlayService.sessionToken
        if (mSessionToken == null && freshToken != null ||
                mSessionToken != null && !(mSessionToken?.equals(freshToken) as Boolean)) {
            if (mController != null) {
                mController?.unregisterCallback(mCb)
            }
            mSessionToken = freshToken
            if (mSessionToken != null) {
                mController = MediaControllerCompat(mPlayService, mSessionToken)
                mTransportControls = mController?.transportControls
                if (mStarted)
                    mController?.registerCallback(mCb)
            }
        }
    }

    private fun createContentIntent(description: MediaDescriptionCompat?): PendingIntent {
        val openUI: Intent = Intent(mPlayService, DetailActivity::class.java)
        openUI.flags = Intent.FLAG_ACTIVITY_SINGLE_TOP
        // todo
        if (description != null) {
            // todo
        }
        return PendingIntent.getActivity(mPlayService, REQUEST_CODE, openUI, PendingIntent.FLAG_CANCEL_CURRENT)
    }

    private val mCb: MediaControllerCompat.Callback = object : MediaControllerCompat.Callback() {
        override fun onPlaybackStateChanged(state: PlaybackStateCompat) {
            mPlaybackState = state
            LogHelper.d(TAG, "Received new playback state", state)
            if (state.state == PlaybackStateCompat.STATE_STOPPED ||
                    state.state == PlaybackStateCompat.STATE_NONE) {
                stopNotification()
            } else {
                val notification: Notification? = createNotification()
                if (notification != null) {
                    mNotificationManager.notify(NOTIFICATION_ID, notification)
                }
            }
        }

        override fun onMetadataChanged(metadata: MediaMetadataCompat?) {
            mMetadata = metadata
            LogHelper.d(TAG, "Received new metadata ", metadata)
            val notification: Notification? = createNotification()
            if (notification != null) {
                mNotificationManager.notify(NOTIFICATION_ID, notification)
            }
        }

        override fun onSessionDestroyed() {
            super.onSessionDestroyed()
            LogHelper.d(TAG, "Session was destroyed, resetting to the new session token")
            try {
                updateSessionToken()
            } catch (e: RemoteException) {
                LogHelper.e(TAG, e, "could not connect media controller")
            }

        }
    }

    private fun createNotification(): Notification? {
        LogHelper.d(TAG, "updateNotificationMetadata. mMetadata=" + mMetadata)
        if (mMetadata == null || mPlaybackState == null)
            return null
        val builder: NotificationCompat.Builder = NotificationCompat.Builder(mPlayService)
        var playPauseButtonPosition: Int = 0

        if ((mPlaybackState?.actions as Long and PlaybackStateCompat.ACTION_SKIP_TO_PREVIOUS) != 0L) {
            builder.addAction(android.R.drawable.ic_media_previous, "上一首", mPreviousIntent)
            playPauseButtonPosition = 1
        }

        addPlayPauseAction(builder)

        if ((mPlaybackState?.actions as Long and PlaybackStateCompat.ACTION_SKIP_TO_NEXT) != 0L) {
            builder.addAction(android.R.drawable.ic_media_next, "下一首", mNextIntent)
        }

        val description: MediaDescriptionCompat = mMetadata?.description as MediaDescriptionCompat
        val art: Bitmap = BitmapFactory.decodeResource(mPlayService.resources, R.drawable.kexuetuokouxiu_2014)

        builder
                .setStyle(NotificationCompat.MediaStyle()
                        .setShowActionsInCompactView(playPauseButtonPosition)
                        .setMediaSession(mSessionToken))
                .setColor(mNotificationColor)
                .setSmallIcon(R.drawable.kexuetuokouxiu_2014)
                .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
                .setUsesChronometer(true)
                .setContentIntent(createContentIntent(description))
                .setContentTitle(description.title)
                .setContentText(description.subtitle)
                .setLargeIcon(art)

        if (mController != null && mController?.extras != null) {
            val castName = mController?.extras?.getString(PlayService.EXTRA_CONNECTED_CAST)
            if (castName != null) {
                val castInfo = mPlayService.resources.getString(R.string.casting_to_device, castName)
                builder.setSubText(castInfo)
                builder.addAction(android.R.drawable.ic_menu_close_clear_cancel, "停止", mStopCastIntent)
            }
        }

        setNotificationPlaybackState(builder)
        return builder.build()
    }

    private fun addPlayPauseAction(builder: NotificationCompat.Builder) {
        LogHelper.d(TAG, "addPlayPauseAction")
        val label: String
        val icon: Int
        val intent: PendingIntent
        if (mPlaybackState?.state == PlaybackStateCompat.STATE_PLAYING) {
            label = "暂停"
            icon = android.R.drawable.ic_media_pause
            intent = mPauseIntent
        } else {
            label = "播放"
            icon = android.R.drawable.ic_media_play
            intent = mPlayIntent
        }
        builder.addAction(android.support.v4.app.NotificationCompat.Action(icon, label, intent))
    }

    private fun setNotificationPlaybackState(builder: NotificationCompat.Builder) {
        LogHelper.d(TAG, "updateNotificationPlaybackState. mPlaybackState=" + mPlaybackState)
        if (mPlaybackState == null || !mStarted) {
            LogHelper.d(TAG, "updateNotificationPlaybackState. cancelling notification!")
            mPlayService.stopForeground(true)
            return
        }
        if (mPlaybackState?.state == PlaybackStateCompat.STATE_PLAYING
                && mPlaybackState?.position as Long >= 0) {
            LogHelper.d(TAG, "updateNotificationPlaybackState. updating playback position to ",
                    (System.currentTimeMillis() - mPlaybackState?.position as Long) / 1000, " seconds")
            builder
                    .setWhen(System.currentTimeMillis() - mPlaybackState?.position as Long)
                    .setShowWhen(true)
                    .setUsesChronometer(true)
        } else {
            LogHelper.d(TAG, "updateNotificationPlaybackState. hiding playback position")
            builder
                    .setWhen(0)
                    .setShowWhen(false)
                    .setUsesChronometer(false)
        }
        // 确保在不播放的时候通知可被清除
        builder.setOngoing(mPlaybackState?.state == PlaybackStateCompat.STATE_PLAYING)
    }
}